src/v/kafka/server/group_metadata.cc

```diff
@@ -390,11 +392,31 @@ group_metadata_serializer make_backward_compatible_serializer() {
     };
     return group_metadata_serializer(std::make_unique<impl>());
 }
+iobuf maybe_unwrap_from_iobuf(iobuf buffer) {
+    /*
+     * Previously in redpanda we had an error leading to wrapping
+     * tombstone record keys with an iobuf serialization envelope. In order to
+     * provide compatibility and be able to read keys wrapped with an additional
+     * iobuf we introduce this check. The check validates the key version being
+     * in allowed range. otherwise we try to unwrap a key from iobuf.
+     *
+     * TODO: remove this code in future releases
+     *
+     */
+    iobuf_parser parser(buffer.copy());
+    int16_t key_version = request_reader(parser.peek(2)).read_int16();
+    if (unlikely(key_version < 0 || key_version > 2)) {
```

Comment:
I'm curious where these versions are coming from. Are these envelope versions, or some storage internal version? It's a little surprising to see that this is checking for `key_version >` since you'd typically expect higher versions to be the most up-to-date. That said, I vaguely recall hearing discussions of negative versions being used to hack around some compatibility issues, so I suspect that's what's being referred to here? Could you add a comment clarifying the context regarding versioning either way?

Reply:
The version is coming from Kafka. Recently we introduced a `__consumer_offsets` topic which contains metadata that are stored in the same format as Kafka uses. (some of the available tools consume that topic to provide insight into consumer groups state) In Kafka they have a notion of a `version`. Here they decided to use version to encode not the version information itself but information about the type of metadata that are stored in record value field.